### PR TITLE
HttpUtils: set Method to HEAD for HttpUtils.getHeaders

### PR DESCRIPTION
**Pull request by lindenb**
### Description

Minor change: in httpUtils `getHeaderField` I set the Http-method to '**HEAD**'  instead of '**GET**' in order to only fetch the header fields.

I also added a test file for HttpUtils.

### Checklist

- [x] Code compiles correctly
- [x] New tests covering changes and new functionality
- [x] All tests passing
- [x] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

### CODE REVIEW

File: src/main/java/htsjdk/samtools/util/HttpUtils.java [hunk 1]
```diff
@@ -29,9 +29,14 @@ public static String getHeaderField(final URL url, final String name) {
         try {
             // Create a URLConnection object for a URL
             conn = openConnection(url);
+            if(conn instanceof HttpURLConnection) {
```
Comment (pshapiro4broad) [inline src/main/java/htsjdk/samtools/util/HttpUtils.java]: need space after `if`

File: src/test/java/htsjdk/samtools/util/HttpUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import java.net.URL;
+
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import htsjdk.HtsjdkTest;
+
+public class HttpUtilsTest extends HtsjdkTest {
+    @DataProvider(name = "existing_urls")
```
Comment (pshapiro4broad) [inline src/test/java/htsjdk/samtools/util/HttpUtilsTest.java]: If you only have a single test case, it would be easier to create a static field of type `URL` than use a data provider.

File: src/test/java/htsjdk/samtools/util/HttpUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    public Object[][] testExistingURLsData() {
+        return new Object[][]{
+                {"http://broadinstitute.github.io/picard/testdata/index_test.bam"}
+        };
+    }
+    
+    @Test(dataProvider="existing_urls")
+    public void testGetHeaderField(final String url) throws IOException {
+    final String field = HttpUtils.getHeaderField(
```
Comment (pshapiro4broad) [inline src/test/java/htsjdk/samtools/util/HttpUtilsTest.java]: Indent is wrong here, it would be best to reformat the entire file since there are many lines with incorrect indent.

File: src/test/java/htsjdk/samtools/util/HttpUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            "Content-Length"
+            );
+    Assert.assertNotNull(field);
+    final long length = Long.parseLong(field);
+    Assert.assertTrue(length>0L);
+    }
+    
+    @Test(dataProvider="existing_urls")
+    public void testGetETag(final String url) throws IOException {
+    final String field = HttpUtils.getETag(new URL(url));
```
Comment (pshapiro4broad) [inline src/test/java/htsjdk/samtools/util/HttpUtilsTest.java]: Is this a useful API to keep? It's not used in picard or htsjdk or gatk. It would be better to remove/deprecate `getEtag()` and `resourceAvailable` than add tests for them.

Comment (lbergelson) [reply src/test/java/htsjdk/samtools/util/HttpUtilsTest.java]: You never know who's using something downstream.  I think we're hoping to deprecate the entire http code package in the future once we've tested out an http path replacement, but I'm not sure if it's worth deprecating bits now.

File: src/main/java/htsjdk/samtools/util/HttpUtils.java [hunk 5]
```diff
@@ -29,9 +29,14 @@ public static String getHeaderField(final URL url, final String name) {
...
+                // The HEAD method is identical to GET except that the server MUST NOT return a message-body in the response.
+                final HttpURLConnection httpConn = HttpURLConnection.class.cast(conn);
+                httpConn.setRequestMethod("HEAD");
+            }
             return conn.getHeaderField(name);
 
-        } catch (IOException e) {
+        } catch (final IOException e) {
             e.printStackTrace();
```
Comment (lbergelson) [inline src/main/java/htsjdk/samtools/util/HttpUtils.java]: We should probably be crashing here instead, but lets leave that for another pr.

File: src/main/java/htsjdk/samtools/util/HttpUtils.java [hunk 6]
```diff
@@ -29,9 +29,14 @@ public static String getHeaderField(final URL url, final String name) {
         try {
             // Create a URLConnection object for a URL
             conn = openConnection(url);
+            if(conn instanceof HttpURLConnection) {
+                // The HEAD method is identical to GET except that the server MUST NOT return a message-body in the response.
+                final HttpURLConnection httpConn = HttpURLConnection.class.cast(conn);
```
Comment (lbergelson) [inline src/main/java/htsjdk/samtools/util/HttpUtils.java]: Why not just `(HttpUrlConnection) conn`?

**Comment by lindenb** [comment]
@lbergelson 

> Do most major web servers support HEAD as well as they support GET?

citing  John Marshall on twitter: https://twitter.com/jomarnz/status/1049762552304074752

> RFC 7231 is definitive on this and says “All general-purpose servers MUST support the methods GET and HEAD”. Me, I test stuff with curl -I (which selects HEAD) all the time.

**Comment by lindenb** [comment]
@lbergelson 
 
>  is it actually a noticeable improvement? 

a 'stupid test':

```

$ time ( for i in {1..500}; do  curl -s  -I -D - "http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/current.tree"   > /dev/null ; done)

real	0m59.979s
user	0m3.380s
sys	0m2.084s

$ time ( for i in {1..500}; do  curl -s  -D - "http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/current.tree" | grep "ftp" -m1  > /dev/null ; done)

real	1m16.346s
user	0m4.040s
sys	0m3.724s

```

**Comment by lindenb** [comment]
@pshapiro4broad back to you. I kept 'getETag' for another PR

File: src/test/java/htsjdk/samtools/util/HttpUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    public Object[][] testExistingURLsData() {
+        return new Object[][]{
+            {"http://broadinstitute.github.io/picard/testdata/index_test.bam"},
+            {"http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/current.tree"}
+        };
+    }
+
+    @Test(dataProvider="existing_urls")
+    public void testGetHeaderField(final String url) throws IOException {
+        final String field = HttpUtils.getHeaderField(new URL(url), "Content-Length");
```
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/test/java/htsjdk/samtools/util/HttpUtilsTest.java]: I don't think that it will test the right thing if we use HTTPHelper....

File: src/test/java/htsjdk/samtools/util/HttpUtilsTest.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        Assert.assertNotNull(field);
+        final long length = Long.parseLong(field);
+        Assert.assertTrue(length > 0L);
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/HttpUtilsTest.java]: can we assert the actual length or is that not stable?

File: src/test/java/htsjdk/samtools/util/HttpUtilsTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (lbergelson) [reply src/test/java/htsjdk/samtools/util/HttpUtilsTest.java]: Since this test is testing the call to getHeaderField it seems like it would be problematic to replace it with a different method.  You could certainly make a case for merging the two methods but I'm not sure that's necessarily in the scope of this.
